Create yml task-definition files for C tasks.

lembergerth opened the pull request:
Preprocessed files (.i) keep their suffix in the yml file name to differentiate them from the .c files.

The yml files contain the properties that are encoded in the tasks' file names,
as well as coverage properties, based on the following two rules:

1. If a task has as an expected verdict `false-unreach-call`, it always gets the `coverage-error-call` property, as well.
2. If a task is part of any of the `*-testable.set` files, it also gets all other coverage properties (i.e., branch, condition and statement coverage)

Rule 1 means that a task gets the property coverage-error-call, even if it is not part of a testable set.
@dbeyer Should we change this so that only testable tasks get the property?

In addition, C files were not renamed yet.
For most tasks, this is straightforward - I would simply remove the verdict-property parts of each program, e.g.: `Mono6_false-unreach-call1.c` -> `Mono6_1.c`
But there are some task names that only differ in the verdict, e.g.:
```
busybox-1.22.0/sleep_false-no-overflow.i
busybox-1.22.0/sleep_false-valid-deref.i
busybox-1.22.0/sleep_true-no-overflow_true-valid-memsafety.i
```
How should we rename these?

Edit: Things left to do:

  - [x] Make sure (coverage) properties are the way we want them
  - [x] Rename tasks
  - [x] Replace C files in Benchmark set files with yml files
  - [x] Adjust CI Checks

lembergerth commented:
Thanks for your feedback! I simplified the files.

Regarding renaming, I didn't choose the worst case of conflicts.

How about files that only differ in the verdict of the same property, like the following?
```
./array-examples/data_structures_set_multi_proc_false-unreach-call_ground.i
./array-examples/data_structures_set_multi_proc_true-unreach-call_ground.i
```
If we only keep the `true`/`false` part, this could get confusing because of other properties with a different value, so the only "solution" that comes to my mind is to
(a) either keep the names the way they are, or
(b) add a unique number as suffix.

PhilippWendler commented:
For Java tasks, I used `_true`/`_false`, but I would be happy with (a) or (b), too. (a) would be the simplest.
We would rename the source files and the Yaml files, right?

Another suggestion:
Strip the suffix `.i` when creating the Yaml file name, like you do for `.c`. `bAnd4_true-unreach-call.i.yml` would then be named `bAnd4.yml`.
If we ever drop the requirements for preprocessed files, this would be better.

lembergerth commented:
> We would rename the source files and the Yaml files, right?

Yes.

> Strip the suffix .i when creating the Yaml file name

I forgot to explain this properly in the pull request,
but the reason for keeping the suffix `.i` was that there's  a significant amount of tasks for which the .c and the .i files exist.
I decided to remove the `.c` and keep the `.i` (and not the other way around) because the .c-Files are the `default`, from my point of view.
Should I instead just ignore the `.c`-Files whenever both versions exist?

lembergerth commented:
> We would rename the source files and the Yaml files, right?

Yes.

> Strip the suffix .i when creating the Yaml file name

I forgot to explain this properly in the pull request,
but the reason for keeping the suffix `.i` was that there's  a significant amount of tasks for which the .c and the .i files exist.
I decided to remove the `.c` and keep the `.i` (and not the other way around) because the .c-Files are the "default", from my point of view.
Should I instead just ignore the `.c`-Files whenever both versions exist?

PhilippWendler commented:
Are you sure there are *tasks* (as defined in the `.set` files) where both `.i` and `.c` variants exist?

The PR shouldn't create Yaml files for each `.c` and `.i` file in the repo, it should only create Yaml files for each task defined by the sets.

lembergerth commented:
> Are you sure there are tasks (as defined in the .set files) where both .i and .c variants exist?

No, I used all .c and .i files before - I've updated the PR to only create yml-files for actual tasks.
I also performed the rename as documented here - also not considered before, if a file has the ending `.c.i`, it is also omitted.

At the moment, the set files contain lists of all .yml-files, which makes them grow quite big.
Is it safe to introduce wildcards, here?

For a bit more overview, I kept the groups that were implied by the wildcards before
and added them as comments to the sets, preceding each group.
For example:

```
pthread/*_false-unreach-call*.i

-> 


## pthread/*_false-unreach-call*.i
pthread/bigshot_p.yml
pthread/fib_bench_false-unreach-call.yml
pthread/fib_bench_longer_false-unreach-call.yml
...
```

Let me know what you think.

PhilippWendler commented:
> At the moment, the set files contain lists of all .yml-files, which makes them grow quite big.
> Is it safe to introduce wildcards here?

Yes, and it would be preferred. With wildcards we can't group into `true`/`false` tasks anymore, but we decided to accept this for now. Later such a sorting order could be added to BenchExec, if it is important enough.

Could you then please rebase your PR commits such that the intermediate Yaml files go away?

lembergerth commented:
I pushed the wildcarding and an improved rename
that also renames .c files that are next to a preprocessed .i task file.

The wildcarding is less than perfect, but I'm not sure how to improve this - often, simple `directory/*.yml` wildcards are not possible.

PhilippWendler commented:
Actually, `directory/*.yml` wildcards would be possible even for cases like `MemSafety-Other`:
If `MemSafety-Other.set` is used together with `<propertyfile>valid-memsafety.yml</propertyfile>`, BenchExec would filter out all tasks that do not have `valid-memsafety.yml` listed as property.
Advantage would be that new tasks with this property in the existing directories would automatically get picked up.
Disadvantage would be that it would be a little bit harder to understand this and use the sets with other scripts. But probably there are such cases anyway already.

lembergerth commented:
> Advantage would be that new tasks with this property in the existing directories would automatically get picked up.

And humans quickly see by the name of the .set file and the wildcards from which directories tasks with the corresponding property are picked up.

> Disadvantage would be that it would be a little bit harder to understand this and use the sets with other scripts. But probably there are such cases anyway already.

Previously, one had to read the .set files and evaluate the glob-expressions.
Now, one has to, in addition, grep the resulting files for "properties/property_file.prp" - I think, anyone who can do the first will be able to achieve the second :wink:

tautschnig commented:
From your commit message:
> If a task is a preprocessed C file and the original file exists next to it, it is also renamed.

Would you mind going one step further and follow the same rules that the compare script follows (https://github.com/sosy-lab/sv-benchmarks/blob/b0572078e0ccf4a6fc75d5d8afbeb418f63c9de0/c/compare.sh#L84), which also handles a few extra cases?

lembergerth commented:
While adjusting the CI checks, I got to realize three additional things:

1. there are two directories with currently unused tasks, i.e., `regression` and `ldv-multiproperty`. At the moment, yml-files are not yet created for these - but I will do that.
2. There is a property `alloca` encoded in the tasks for which no property file exists. At the moment, this property/verdict is kept in the file name and is not added to the yml-file. I don't plan to change that.
3. There are `unknown` verdicts encoded in some tasks. Same applies as for 2, and I don't plan to change that.

PhilippWendler commented:
`regression` was added but never added to a category. I don't think it is worth adding task-definition files.
`ldv-multiproperty` is for multi-property verification, where other properties will be used. This could be added later, when we have support for multi-property verification and a category is created.

For verdict `unknown` we could also add the respective property to the task-definition file without `expected_verdict`. This would show the intended semantics of "property is meaningful for this program, but verdict is unknown".

dbeyer commented:
@PhilippWendler and @tautschnig: Do you see any issues that we need to solve before merging?

dbeyer commented:
> ./array-examples/data_structures_set_multi_proc_false-unreach-call_ground.i
> ./array-examples/data_structures_set_multi_proc_true-unreach-call_ground.i
> ```
> 
> If we only keep the `true`/`false` part, this could get confusing because of other properties with a different value, so the only solution that comes to my mind is to
> (a) either keep the names the way they are, or
> (b) add a unique number as suffix.

I would slightly favor (b), in order to consistently remove the expected results from the file names.

dbeyer commented:
Example:
```
trex01_false-unreach-call_true-termination.yml  => trex01a.yml
trex01_true-unreach-call_true-termination.yml   => trex01b.yml
trex02_false-unreach-call_true-termination.yml  => trex02a.yml
trex02_true-unreach-call_true-termination.yml   => trex02b.yml
trex03_false-unreach-call_true-termination.yml  => trex03a.yml
trex03_true-unreach-call_true-termination.yml   => trex03b.yml
trex04.yml                                                           => trex04.yml
```

PhilippWendler commented:
@lembergerth Are you going to add a PR that renames the remaining tasks to remove the expected verdicts?

lembergerth commented:
Yes, thanks for the reminder.
